typing/typecore.ml

```diff
@@ -2971,12 +3056,44 @@ and type_expect_
         if rec_flag = Recursive then In_rec
         else if List.compare_length_with spat_sexp_list 1 > 0 then In_group
         else With_attributes in
-      let (pat_exp_list, new_env, unpacks) =
-        type_let existential_context env rec_flag spat_sexp_list true in
-      let body = type_unpacks new_env unpacks sbody ty_expected_explained in
-      let () =
-        if rec_flag = Recursive then
-          check_recursive_bindings env pat_exp_list
+      let may_contain_modules =
+        List.exists (fun pvb -> may_contain_modules pvb.pvb_pat) spat_sexp_list
+      in
+      let (pat_exp_list, body, _new_env) =
+        with_local_level_if may_contain_modules begin fun () ->
+          let allow_modules =
+            if may_contain_modules
+            then
+              let scope = create_scope () in
+              Modules_allowed { scope }
+            else Modules_rejected
+          in
+          let (pat_exp_list, new_env) =
+            type_let existential_context env rec_flag spat_sexp_list
+              allow_modules
+          in
+          let body = type_expect new_env sbody ty_expected_explained in
+          let () =
+            if rec_flag = Recursive then
+              check_recursive_bindings env pat_exp_list
+          in
+          (pat_exp_list, body, new_env)
+        end
+        (* If the patterns contain module unpacks, there is a possibility that
+           the type of the let body or variables bound by the let mention types
+           introduced by those unpacks. (The latter can only happen with
+           recursive definitions.) Here, we check for scope escape via both of
+           these pathways (body, variables).
+
+           Checking unification within an environment extended with the module
+           bindings allows us to correctly accept more programs. This
+           environment allows unification to identify more cases where a type
+           introduced by the module is equal to a type introduced at an outer
+           scope. *)
+        ~post:(fun (pat_exp_list, body, new_env) ->
+          unify_exp new_env body (newvar ());
+          if rec_flag = Recursive then
+            check_scope_escape_let_bound_idents new_env pat_exp_list)
```

Comment:
Yes, I believe this makes the change you described. This check previously happened as part of `type_unpacks`, so now we need to implement something here &mdash; in discussing with @goldfirere, we concluded that checking escape for let-bound idents was all that was necessary.

I don't foresee difficulties if we wanted to instead prohibit types from escaping from the body even if they aren't bound in the pattern. This seems perhaps better as it's closer to what we did before (even if I haven't yet been able to construct any programs that would be accepted or rejected differently according to this change in implementation).

(You are likely aware of this, but just to state this explicitly, this code is only relevant for recursive let-bindings that bind modules, like `let rec (module M) = ...`.)

Reply:
(I just pushed a change to revert to the old approach &mdash; check `E1` in `let P = E1 in E2` instead of just the bound vars in `P` &mdash; and think it feels a bit cleaner.)

Reply:
I did not doubt the correctness of your approach, but the old one is indeed easier to justify, as it can be understood as an inlined `let module`.

Reply:
Yes, I'm happy with the easier-to-justify approach. I'm marking this as resolved because it seems like we're in agreement.